### PR TITLE
Issue 6067 - Improve dsidm CLI No Such Entry handling

### PR DESCRIPTION
**Pull request by droideck**
Add additional error processing to dsidm CLI tool for when basedn or OU subentries are absent.

Related: https://github.com/389ds/389-ds-base/issues/6067

Reviewed by: ?

### CODE REVIEW

File: src/lib389/lib389/cli_idm/__init__.py [hunk 1]
```diff
@@ -1,15 +1,30 @@
...
 from getpass import getpass
 import json
+from lib389._mapped_object import DSLdapObject
+from lib389.cli_base import _get_dn_arg
+from lib389.idm.user import DEFAULT_BASEDN_RDN as DEFAULT_BASEDN_RDN_USER
+from lib389.idm.group import DEFAULT_BASEDN_RDN as DEFAULT_BASEDN_RDN_GROUP
+from lib389.idm.posixgroup import DEFAULT_BASEDN_RDN as DEFAULT_BASEDN_RDN_POSIXGROUP
+from lib389.idm.services import DEFAULT_BASEDN_RDN as DEFAULT_BASEDN_RDN_SERVICES
+
+# Create a dict where key is module and value is an rpm to search
```
Comment (vashirov) [inline src/lib389/lib389/cli_idm/__init__.py]: What does this comment mean?

**Review by vashirov — APPROVED** [review]
It's still a bit awkward that it allows to enter some details only to fail later:
```
dsidm localhost -b dc=example,dc=com service create
Enter value for cn : test
Enter value for description : test
Error: The base DN "ou=Services,dc=example,dc=com" does not exist
```

But the new error message is an improvement :)
LGTM

File: src/lib389/lib389/cli_idm/__init__.py (hunk 1 above)
Comment (droideck) [reply src/lib389/lib389/cli_idm/__init__.py]: I guess doing many builds at that time affected my brain somehow (hence rdn->rpm typo). Fixed.

**Comment by droideck** [comment]
> It's still a bit awkward that it allows to enter some details only to fail later:
> 
> ```
> dsidm localhost -b dc=example,dc=com service create
> Enter value for cn : test
> Enter value for description : test
> Error: The base DN "ou=Services,dc=example,dc=com" does not exist
> ```
> 
> But the new error message is an improvement :) LGTM

Okay, it was actually another typo in the same code area (`services` instead of `service`).
Now it should correctly display `Error: The DN "ou=Services,dc=example,dc=com" does not exist. It is required for "service" subcommand. Please create it first.` error.

**Comment by Firstyear** [comment]
Sorry to re-open this one, but shouldn't https://github.com/389ds/389-ds-base/blob/83668192bddcb6daa272eb418a96c10931d3386b/src/lib389/lib389/cli_idm/__init__.py#L55 be reading the basedn from dsrc as well? Else you always need it with -b.

**Comment by progier389** [comment]
Hi @Firstyear .
IMHO, it would change the current behavior:
   If I understood rightly in current code the base dn is interactively asked if -b option is missing (like for other dsidm mandatory options).
   So if we use a default value from dsrc there will be no way to interactively ask for values
   unless we choose to drastically change the interface (for example by adding an interactive mode option)
I have no strong feeling about that (on the plus side: it would make dsidm more like the other dsxxx commands, 
 but on the minus side it will change the interface and may disrupt some customer scripts/habits.)
In any case, if we want to pursue this idea, I think it is unrelated to the original issue and deserves its own ticket.
What do you think ?

**Comment by Firstyear** [comment]
But if the default value is set, it's assumed the admin wants to use that *unless* they override it with -b? 

We had a customer report this as a regression in behaviour, which is why I bring it up :)

**Comment by droideck** [comment]
> But if the default value is set, it's assumed the admin wants to use that _unless_ they override it with -b?
> 
> We had a customer report this as a regression in behaviour, which is why I bring it up :)
> 
> edit: https://bugzilla.suse.com/show_bug.cgi?id=1231462

I agree. I think it should be `[check if -b provided]` -> if not -> `[check if basedn is in .dsrc]` -> if not -> `[ask interactively]`.
I'll prepare PR, it's an easy fix.

**Pull request merged**